src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdCaseUpdater.java

```diff
@@ -0,0 +1,189 @@
+    private final AuthTokenGenerator s2sTokenGenerator;
+    private final CoreCaseDataApi coreCaseDataApi;
+    private final ExceptionRecordFinalizer exceptionRecordFinalizer;
+
+    public CcdCaseUpdater(
+        CaseUpdateClient caseUpdateClient,
+        AuthTokenGenerator s2sTokenGenerator,
+        CoreCaseDataApi coreCaseDataApi,
+        ExceptionRecordFinalizer exceptionRecordFinalizer
+    ) {
+        this.caseUpdateClient = caseUpdateClient;
+        this.s2sTokenGenerator = s2sTokenGenerator;
+        this.coreCaseDataApi = coreCaseDataApi;
+        this.exceptionRecordFinalizer = exceptionRecordFinalizer;
+    }
+
+    @SuppressWarnings({"squid:S2139", "unchecked"}) // squid for exception handle + logging
+    public ProcessResult updateCase(
+        ExceptionRecord exceptionRecord,
+        ServiceConfigItem configItem,
+        boolean ignoreWarnings,
+        String idamToken,
+        String userId,
+        CaseDetails existingCase
+    ) {
+        log.info(
+            "Start updating case for service {} with case ID {} from exception record {}",
+            configItem.getService(),
+            existingCase.getId(),
+            exceptionRecord.id
+        );
+
+        try {
+            String s2sToken = s2sTokenGenerator.generate();
+
+            SuccessfulUpdateResponse updateResponse = caseUpdateClient.updateCase(
+                configItem.getUpdateUrl(),
+                existingCase,
+                exceptionRecord,
+                s2sToken
+            );
+
+            if (!ignoreWarnings && !updateResponse.warnings.isEmpty()) {
+                // do not log warnings
+                return new ProcessResult(updateResponse.warnings, emptyList());
+            }
+
+            log.info(
+                "Successfully called case update endpoint of service {} to update case with case ID {} "
+                    + "based on exception record ref {}",
+                configItem.getService(),
+                existingCase.getId(),
+                exceptionRecord.id
+            );
+
+            updateCaseInCcd(
+                ignoreWarnings,
+                idamToken,
+                s2sToken,
+                userId,
+                exceptionRecord,
+                updateResponse.caseDetails,
+                existingCase
+            );
+
+            log.info(
+                "Successfully updated case for service {} with case ID {} based on exception record ref {}",
+                configItem.getService(),
+                existingCase.getId(),
+                exceptionRecord.id
+            );
+
+            return new ProcessResult(
+                exceptionRecordFinalizer.finalizeExceptionRecord(existingCase.getData(), existingCase.getId())
+            );
+        } catch (InvalidCaseDataException exception) {
+            if (BAD_REQUEST.equals(exception.getStatus())) {
+                throw new CallbackException("Failed to update case", exception);
```

Comment:
The message for this case should be more like `Case Update API returned an error response with status {}.` Plus service name, exception record id and case ID

Reply:
Will change